src/wagtail_personalisation/models.py

```diff
@@ -107,6 +149,93 @@ def toggle(self, save=True):
             self.save()
 
 
+class SegmentVisitMetadata(models.Model):
+    visit = models.ForeignKey(
+        'wagtail_personalisation.SegmentVisit', on_delete=models.CASCADE)
+    segment = models.ForeignKey(
+        'wagtail_personalisation.Segment', on_delete=models.SET_NULL, null=True)
+    matched_rules = models.CharField(max_length=2048)
+
+
+class SegmentVisit(models.Model):
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL, on_delete=models.SET_NULL, null=True)
+    page = models.ForeignKey(Page, on_delete=models.SET_NULL, null=True)
+    segments = models.ManyToManyField(Segment, through=SegmentVisitMetadata)
+    served_segment = models.ForeignKey(
+        Segment, on_delete=models.CASCADE,
+        related_name='served_segment', null=True)
+    served_variant = models.ForeignKey(
+        Page, on_delete=models.SET_NULL,
+        related_name='served_variant', null=True)
+    session = models.CharField(
+        max_length=64, editable=False, null=True, db_index=True)
+    visit_date = models.DateTimeField(auto_now_add=True)
+
+    @classmethod
+    def create_segment_visit(cls, page, request, metadata=None):
+        """Create a segment visit object.
+        :param page: The page being visited
+        :type page: wagtail.wagtailcore.models.Page
+        :param request: The http request
+        :type request: django.http.HttpRequest
+        :param metadata: A list of personalisable page metadata
+        :type page: wagtail_personalisation.models.PersonalisablePageMetadata
+        :returns: A committed Segment Visit object
+        :rtype: wagtail_personalisation.models.SegmentVisit
+        """
+        from wagtail_personalisation.adapters import get_segment_adapter
+
+        adapter = get_segment_adapter(request)
+        user_segments = adapter.get_segments()
+
+        if not metadata:
+            metadata = page.personalisation_metadata
+            metadata = metadata.metadata_for_segments(user_segments)
+
+        user = request.user if request.user.is_authenticated() else None
+        visit = cls.objects.create(
+            user=user,
+            page=page,
+            served_segment=metadata.first().segment,
+            served_variant=metadata.first().variant,
+            session=request.session.session_key
+        )
+
+        for segment in user_segments:
+            rules = [
+                rule for rule in segment.get_rules() if rule.unique_encoded_name
+                in request.matched_rules
+            ]
+
+            SegmentVisitMetadata.objects.create(
+                visit=visit,
+                segment=segment,
+                matched_rules=','.join(
+                    rule.unique_encoded_name for rule in rules)
+            )
+
+        return visit
+
+    @classmethod
+    def reverse_match(cls, user):
```

Comment:
The `session_key` used in the `SegmentVisit` class changes every time a user logs in and logs out. Is there an other way to recognise previous visits of a user without using this session key?

Reply:
Might be the secret key not being set properly.